### PR TITLE
Plugin wizard helpers

### PR DESCRIPTION
**Pull request by tfennelly**
Some changes to help with Plugin Wizard (and plugin unbundling) changes for Jenkins 2.0.

- [x] Migrate changes from old Jenkins 2.0 branch.
- [x] Test against 2.0 branch (make sure it does what we need for 2.0).
- [x] Test against 1.x branch (make sure if's backward compatible).

### CODE REVIEW

File: src/main/java/org/jvnet/hudson/test/TestPluginManager.java [hunk 1]
```diff
@@ -63,11 +66,24 @@ private TestPluginManager() throws IOException {
...
+        
+        names.addAll(loadBundledPlugins(new File(WarExploder.getExplodedDir(), "WEB-INF/plugins")));
+        loadBundledPlugins(new File(WarExploder.getExplodedDir(), "WEB-INF/detached-plugins"));
+                
+        return names;
+    }
+
+    private Set<String> loadBundledPlugins(File fromDir) throws IOException, URISyntaxException {
+        if (!fromDir.exists()) {
+            LOGGER.log(Level.INFO, "No plugins loaded from " + fromDir + ". Directory doesn't exist.");
```
Comment (jglick) [inline src/main/java/org/jvnet/hudson/test/TestPluginManager.java]: Maybe shut this up for the common case that you are running against 1.0 and `detached-plugins` is not _supposed_ to exist.

File: src/main/java/org/jvnet/hudson/test/TestPluginManager.java [hunk 2]
```diff
@@ -113,6 +129,19 @@ private TestPluginManager() throws IOException {
         return names;
     }
     
+    /**
+     * Install a plugin from the resources directory.
+     * @param pluginName The plugin name.
+     * @throws IOException Error copying plugin.
+     */
+    public void installResourcePlugin(String pluginName) throws Exception {
+        URL res = getClass().getClassLoader().getResource("plugins/" + pluginName);
```
Comment (jglick) [inline src/main/java/org/jvnet/hudson/test/TestPluginManager.java]: `getClass().getClassLoader().getResource(…)` is generally an antipattern. What is this `Class`? Might it have been subclassed in a different package? If you are sure that the class loader you want is that which loaded `TestPluginManager`, then say so: `TestPluginManager.class.getClassLoader().getResource(…)`.

**Comment by jglick** [comment]
Looks OK so far.

**Comment by tfennelly** [comment]
@reviewbybees

File: pom.xml [hunk 3]
```diff
@@ -33,7 +33,7 @@ THE SOFTWARE.
 
   <groupId>org.jenkins-ci.main</groupId>
   <artifactId>jenkins-test-harness</artifactId>
-  <version>2.1-SNAPSHOT</version>
+  <version>2.1-beta-1</version>
```
Comment (jglick) [inline pom.xml]: Why?

File: src/main/java/org/jvnet/hudson/test/JenkinsRule.java [hunk 4]
```diff
@@ -913,6 +922,114 @@ public Page search(String q) throws Exception {
     }
 
     /**
+     * Get JSON from A Jenkins endpoint.
+     * @param path The endpoint URL.
+     * @return The JSON.
+     */
+    public JSONWebResponse getJSON(@Nonnull String path) throws IOException {
```
Comment (olivergondza) [inline src/main/java/org/jvnet/hudson/test/JenkinsRule.java]: This is unused and it does not belong to the rule class.

File: src/main/java/org/jvnet/hudson/test/JenkinsRule.java [hunk 5]
```diff
@@ -913,6 +922,114 @@ public Page search(String q) throws Exception {
...
+
+    /**
+     * POST a JSON payload to a URL on the underlying Jenkins instance.
+     * @param path The url path on Jenkins.
+     * @param json An object that produces a JSON string from it's {@code toString} method.
+     * @return A JSON response.
+     * @throws IOException
+     * @throws SAXException
+     */
+    public JSONWebResponse postJSON(@Nonnull String path, @Nonnull Object json) throws IOException, SAXException {
```
Comment (olivergondza) [inline src/main/java/org/jvnet/hudson/test/JenkinsRule.java]: This is unused and it does not belong to the rule class.

**Comment by jglick** [comment]
Merge with #3 to kick off a CI build.

**Comment by tfennelly** [comment]
@jglick merged ... hopefully we can merge this and release ASAP. It's blocking [jenkins #1995](https://github.com/jenkinsci/jenkins/issues/1995).

**Comment by tfennelly** [comment]
@jenkinsci/code-reviewers @reviewbybees

File: src/main/java/org/jvnet/hudson/test/TestPluginManager.java [hunk 6]
```diff
@@ -63,11 +66,24 @@ private TestPluginManager() throws IOException {
     @Override
     protected Collection<String> loadBundledPlugins() throws Exception {
         Set<String> names = new HashSet<String>();
+        
+        names.addAll(loadBundledPlugins(new File(WarExploder.getExplodedDir(), "WEB-INF/plugins")));
+        loadBundledPlugins(new File(WarExploder.getExplodedDir(), "WEB-INF/detached-plugins"));
```
Comment (jglick) [inline src/main/java/org/jvnet/hudson/test/TestPluginManager.java]: Intentionally not added to `names`?

File: src/main/java/org/jvnet/hudson/test/TestPluginManager.java [hunk 7]
```diff
@@ -113,6 +129,19 @@ private TestPluginManager() throws IOException {
...
     
+    /**
+     * Install a plugin from the resources directory.
+     * @param pluginName The plugin name.
+     * @throws IOException Error copying plugin.
+     */
+    public void installResourcePlugin(String pluginName) throws Exception {
+        URL res = TestPluginManager.class.getClassLoader().getResource("plugins/" + pluginName);
+        if (res == null) {
+            Assert.fail("Plugin '" + pluginName + "' not found in /resources/plugins.");
```
Comment (jglick) [inline src/main/java/org/jvnet/hudson/test/TestPluginManager.java]: :ant: Why `/resources`? Potentially misleading.

BTW easier to use `assertNotNull`.

**Comment by jglick** [comment]
:bee:

File: src/main/java/org/jvnet/hudson/test/TestPluginManager.java (hunk 6 above)
Comment (tfennelly) [inline src/main/java/org/jvnet/hudson/test/TestPluginManager.java]: Correct... that's intentional.

File: src/main/java/org/jvnet/hudson/test/TestPluginManager.java (hunk 7 above)
Comment (tfennelly) [inline src/main/java/org/jvnet/hudson/test/TestPluginManager.java]: > BTW easier to use `assertNotNull`.

Hmmm... a matter of taste I think.

File: src/main/java/org/jvnet/hudson/test/JenkinsRule.java [hunk 8]
```diff
@@ -913,6 +922,114 @@ public Page search(String q) throws Exception {
...
+        assert !path.startsWith("/");
+
+        URL postUrl = new URL(getURL().toExternalForm() + path);
```
Comment (amuniz) [inline src/main/java/org/jvnet/hudson/test/JenkinsRule.java]: Is `getURL().toExternalForm()` returning always the slash at the end?

File: src/main/java/org/jvnet/hudson/test/JenkinsRule.java [hunk 9]
```diff
@@ -913,6 +922,114 @@ public Page search(String q) throws Exception {
...
+            conn.setRequestProperty(crumb.getName(), crumb.getValue());
+
+            byte[] content = json.toString().getBytes(UTF8);
+            conn.setRequestProperty("Content-Length", String.valueOf(content.length));
+            final OutputStream os = conn.getOutputStream();
+            try {
+                os.write(content);
+                os.flush();
+            } finally {
+                os.close();
```
Comment (amuniz) [inline src/main/java/org/jvnet/hudson/test/JenkinsRule.java]: `IOUtils.closeQuietly`?

File: src/main/java/org/jvnet/hudson/test/JenkinsRule.java [hunk 10]
```diff
@@ -2246,5 +2361,9 @@ public Description getTestDescription() {
         return testDescription;
     }
 
+    private NameValuePair getCrumbHeaderNVP() {
+        return new NameValuePair(jenkins.getCrumbIssuer().getDescriptor().getCrumbRequestField(),
+                        jenkins.getCrumbIssuer().getCrumb( null ));
```
Comment (amuniz) [inline src/main/java/org/jvnet/hudson/test/JenkinsRule.java]: I know it's copied from below, but you are inheriting the sin :smile_cat: (spaces around `null`).

**Comment by amuniz** [comment]
:bee:

File: src/main/java/org/jvnet/hudson/test/JenkinsRule.java (hunk 8 above)
Comment (jglick) [inline src/main/java/org/jvnet/hudson/test/JenkinsRule.java]: [Yes](https://github.com/tfennelly/jenkins-test-harness/blob/dfac492d95060e1ad41c75c425cf09982d0b5c9e/src/main/java/org/jvnet/hudson/test/JenkinsRule.java#L798).

BTW `toString()` is an alias for `toExternalForm()` so `getURL() + path` suffices.

File: src/main/java/org/jvnet/hudson/test/JenkinsRule.java (hunk 9 above)
Comment (jglick) [inline src/main/java/org/jvnet/hudson/test/JenkinsRule.java]: :-1:, simpler and better as written (and when `-source 7` can use a `try`-with-resources block).

File: src/main/java/org/jvnet/hudson/test/JenkinsRule.java (hunk 10 above)
Comment (tfennelly) [inline src/main/java/org/jvnet/hudson/test/JenkinsRule.java]: @amuniz hehehe good job I've no religious beliefs ... I'm easy with having a few sins :wink:

File: src/main/java/org/jvnet/hudson/test/JenkinsRule.java [hunk 11]
```diff
@@ -912,6 +921,114 @@ public Page search(String q) throws Exception {
...
+                os.write(content);
+                os.flush();
+            } finally {
+                os.close();
+            }
+
+            WebResponseData webResponseData;
+            InputStream responseStream = conn.getInputStream();
+            try {
+                if (responseStream != null) {
```
Comment (jglick) [inline src/main/java/org/jvnet/hudson/test/JenkinsRule.java:982]: https://github.com/jenkinsci/jenkins-test-harness/pull/558#discussion_r1120231642

**Pull request merged**